### PR TITLE
pkg/maps: Fix data races around accessing nat maps

### LINKED ISSUE #18910: race: between `pkg/bpf.(*Map).Close()` and `pkg/endpoint.(*Endpoint).runPreCompilationSteps()`
Observed in https://jenkins.cilium.io/view/Race%20Detection/job/cilium-master-gke-race-detection/263/

Artifacts: [573565b7_K8sDatapathConfig_Host_firewall_With_VXLAN.zip](https://github.com/cilium/cilium/files/8125930/573565b7_K8sDatapathConfig_Host_firewall_With_VXLAN.zip)

Trace:
```
2022-02-21T02:42:13.757888057Z WARNING: DATA RACE
2022-02-21T02:42:13.757895621Z Write at 0x00c00081a830 by goroutine 360:
2022-02-21T02:42:13.757912783Z   github.com/cilium/cilium/pkg/bpf.(*Map).Close()
2022-02-21T02:42:13.757920669Z       /go/src/github.com/cilium/cilium/pkg/bpf/map_linux.go:581 +0x1c4
2022-02-21T02:42:13.757927597Z   github.com/cilium/cilium/pkg/maps/ctmap.doGC4·dwrap·4()
2022-02-21T02:42:13.757934561Z       /go/src/github.com/cilium/cilium/pkg/maps/ctmap/ctmap.go:422 +0x39
2022-02-21T02:42:13.757941184Z   github.com/cilium/cilium/pkg/maps/ctmap.doGC4()
2022-02-21T02:42:13.757947352Z       /go/src/github.com/cilium/cilium/pkg/maps/ctmap/ctmap.go:482 +0x4d5
2022-02-21T02:42:13.757953881Z   github.com/cilium/cilium/pkg/maps/ctmap.doGC()
2022-02-21T02:42:13.757960362Z       /go/src/github.com/cilium/cilium/pkg/maps/ctmap/ctmap.go:517 +0x17e
2022-02-21T02:42:13.757966185Z   github.com/cilium/cilium/pkg/maps/ctmap.GC()
2022-02-21T02:42:13.757972307Z       /go/src/github.com/cilium/cilium/pkg/maps/ctmap/ctmap.go:537 +0xc6
2022-02-21T02:42:13.757978599Z   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).garbageCollectConntrack()
2022-02-21T02:42:13.757986321Z       /go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:1034 +0x804
2022-02-21T02:42:13.757992160Z   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).scrubIPsInConntrackTableLocked()
2022-02-21T02:42:13.757998853Z       /go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:1039 +0x173
2022-02-21T02:42:13.758004601Z   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).scrubIPsInConntrackTable()
2022-02-21T02:42:13.758010701Z       /go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:1049 +0x44
2022-02-21T02:42:13.758016604Z   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).runPreCompilationSteps.func1()
2022-02-21T02:42:13.758022804Z       /go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:761 +0x132
2022-02-21T02:42:13.758028469Z 
2022-02-21T02:42:13.758034551Z Previous read at 0x00c00081a830 by goroutine 100:
2022-02-21T02:42:13.758040659Z   github.com/cilium/cilium/pkg/bpf.(*Map).DumpReliablyWithCallback()
2022-02-21T02:42:13.758046461Z       /go/src/github.com/cilium/cilium/pkg/bpf/map_linux.go:756 +0x804
2022-02-21T02:42:13.758053696Z   github.com/cilium/cilium/pkg/maps/nat.(*Map).DumpReliablyWithCallback()
2022-02-21T02:42:13.758059818Z       /go/src/github.com/cilium/cilium/pkg/maps/nat/nat.go:121 +0x7b7
2022-02-21T02:42:13.758065580Z   github.com/cilium/cilium/pkg/maps/ctmap.PurgeOrphanNATEntries()
2022-02-21T02:42:13.758072272Z       /go/src/github.com/cilium/cilium/pkg/maps/ctmap/ctmap.go:612 +0x790
2022-02-21T02:42:13.758078005Z   github.com/cilium/cilium/pkg/maps/ctmap/gc.runGC()
2022-02-21T02:42:13.758084196Z       /go/src/github.com/cilium/cilium/pkg/maps/ctmap/gc/gc.go:214 +0xb0c
2022-02-21T02:42:13.758090362Z   github.com/cilium/cilium/pkg/maps/ctmap/gc.runGC()
2022-02-21T02:42:13.758096712Z       /go/src/github.com/cilium/cilium/pkg/maps/ctmap/gc/gc.go:189 +0x6a5
2022-02-21T02:42:13.758103134Z   github.com/cilium/cilium/pkg/maps/ctmap/gc.runGC()
2022-02-21T02:42:13.758109338Z       /go/src/github.com/cilium/cilium/pkg/maps/ctmap/gc/gc.go:189 +0x6a5
2022-02-21T02:42:13.758127517Z   github.com/cilium/cilium/pkg/maps/ctmap/gc.Enable.func1()
2022-02-21T02:42:13.758135513Z       /go/src/github.com/cilium/cilium/pkg/maps/ctmap/gc/gc.go:90 +0x40a
2022-02-21T02:42:13.758141621Z   github.com/cilium/cilium/pkg/maps/ctmap/gc.runGC()
2022-02-21T02:42:13.758147715Z       /go/src/github.com/cilium/cilium/pkg/maps/ctmap/gc/gc.go:189 +0x6a5
2022-02-21T02:42:13.758153712Z   github.com/cilium/cilium/pkg/maps/ctmap/gc.Enable.func1()
2022-02-21T02:42:13.758160127Z       /go/src/github.com/cilium/cilium/pkg/maps/ctmap/gc/gc.go:90 +0x40a
2022-02-21T02:42:13.758165542Z 
2022-02-21T02:42:13.758172000Z Goroutine 360 (running) created at:
2022-02-21T02:42:13.758180580Z   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).runPreCompilationSteps()
2022-02-21T02:42:13.758186960Z       /go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:752 +0x2a8
2022-02-21T02:42:13.758192708Z   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).regenerateBPF()
2022-02-21T02:42:13.758293641Z       /go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:584 +0x226
2022-02-21T02:42:13.758301429Z   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).regenerate()
2022-02-21T02:42:13.758308453Z       /go/src/github.com/cilium/cilium/pkg/endpoint/policy.go:398 +0xc6a
2022-02-21T02:42:13.758314506Z   github.com/cilium/cilium/pkg/endpoint.(*EndpointRegenerationEvent).Handle()
2022-02-21T02:42:13.758321021Z       /go/src/github.com/cilium/cilium/pkg/endpoint/events.go:53 +0x2f6
2022-02-21T02:42:13.758335582Z   github.com/cilium/cilium/pkg/eventqueue.(*EventQueue).run.func1()
2022-02-21T02:42:13.758343444Z       /go/src/github.com/cilium/cilium/pkg/eventqueue/eventqueue.go:245 +0x206
2022-02-21T02:42:13.758349541Z   sync.(*Once).doSlow()
2022-02-21T02:42:13.758357337Z       /usr/local/go/src/sync/once.go:68 +0x127
2022-02-21T02:42:13.758400068Z   sync.(*Once).Do()
2022-02-21T02:42:13.758406993Z       /usr/local/go/src/sync/once.go:59 +0x46
2022-02-21T02:42:13.758508649Z   github.com/cilium/cilium/pkg/eventqueue.(*EventQueue).run()
2022-02-21T02:42:13.758532273Z       /go/src/github.com/cilium/cilium/pkg/eventqueue/eventqueue.go:233 +0x50
2022-02-21T02:42:13.758540684Z   github.com/cilium/cilium/pkg/eventqueue.(*EventQueue).Run·dwrap·2()
2022-02-21T02:42:13.758553454Z       /go/src/github.com/cilium/cilium/pkg/eventqueue/eventqueue.go:229 +0x39
2022-02-21T02:42:13.758559003Z 
2022-02-21T02:42:13.758565564Z Goroutine 100 (running) created at:
2022-02-21T02:42:13.758572619Z   github.com/cilium/cilium/pkg/maps/ctmap/gc.Enable()
2022-02-21T02:42:13.758579844Z       /go/src/github.com/cilium/cilium/pkg/maps/ctmap/gc/gc.go:40 +0x364
2022-02-21T02:42:13.758586062Z   github.com/cilium/cilium/daemon/cmd.runDaemon()
2022-02-21T02:42:13.758592340Z       /go/src/github.com/cilium/cilium/daemon/cmd/daemon_main.go:1702 +0x1211
2022-02-21T02:42:13.758648613Z   github.com/sirupsen/logrus.(*Entry).Info()
2022-02-21T02:42:13.758710157Z       /go/src/github.com/cilium/cilium/vendor/github.com/sirupsen/logrus/entry.go:310 +0x7125
2022-02-21T02:42:13.758753160Z   github.com/cilium/cilium/daemon/cmd.NewDaemon()
2022-02-21T02:42:13.758760755Z       /go/src/github.com/cilium/cilium/daemon/cmd/daemon.go:929 +0x70e5
2022-02-21T02:42:13.758767368Z   github.com/cilium/cilium/daemon/cmd.NewDaemon()
2022-02-21T02:42:13.758992586Z       /go/src/github.com/cilium/cilium/daemon/cmd/daemon.go:888 +0x6911
2022-02-21T02:42:13.759009051Z   github.com/cilium/cilium/daemon/cmd.NewDaemon()
2022-02-21T02:42:13.759015614Z       /go/src/github.com/cilium/cilium/daemon/cmd/daemon.go:708 +0x4424
2022-02-21T02:42:13.759021802Z   github.com/cilium/cilium/daemon/cmd.runDaemon()
2022-02-21T02:42:13.759028515Z       /go/src/github.com/cilium/cilium/daemon/cmd/daemon_main.go:1680 +0xa98
2022-02-21T02:42:13.759045881Z   github.com/cilium/cilium/daemon/cmd.glob..func1()
2022-02-21T02:42:13.759052809Z       /go/src/github.com/cilium/cilium/daemon/cmd/daemon_main.go:126 +0x494
2022-02-21T02:42:13.759059009Z   github.com/cilium/cilium/daemon/cmd.glob..func1()
2022-02-21T02:42:13.759065205Z       /go/src/github.com/cilium/cilium/daemon/cmd/daemon_main.go:124 +0x47e
2022-02-21T02:42:13.759115100Z   github.com/spf13/cobra.(*Command).execute()
2022-02-21T02:42:13.759124422Z       /go/src/github.com/cilium/cilium/vendor/github.com/spf13/cobra/command.go:860 +0xaf3
2022-02-21T02:42:13.759131527Z   github.com/spf13/cobra.(*Command).ExecuteC()
2022-02-21T02:42:13.759137857Z       /go/src/github.com/cilium/cilium/vendor/github.com/spf13/cobra/command.go:974 +0x5da
2022-02-21T02:42:13.759144153Z   github.com/spf13/cobra.(*Command).Execute()
2022-02-21T02:42:13.759150354Z       /go/src/github.com/cilium/cilium/vendor/github.com/spf13/cobra/command.go:902 +0x65
2022-02-21T02:42:13.759156804Z   github.com/cilium/cilium/daemon/cmd.Execute()
2022-02-21T02:42:13.759163045Z       /go/src/github.com/cilium/cilium/daemon/cmd/daemon_main.go:140 +0x4e
2022-02-21T02:42:13.759169325Z   main.main()
2022-02-21T02:42:13.759175845Z       /go/src/github.com/cilium/cilium/daemon/main.go:16 +0x24
2022-02-21T02:42:13.759182494Z   runtime.main()
2022-02-21T02:42:13.759189008Z       /usr/local/go/src/runtime/proc.go:255 +0x226
2022-02-21T02:42:13.759195139Z   github.com/cilium/ebpf.init()
2022-02-21T02:42:13.759201101Z       /go/src/github.com/cilium/cilium/vendor/github.com/cilium/ebpf/syscalls.go:220 +0x868
2022-02-21T02:42:13.759207757Z   github.com/cilium/ebpf.init()
2022-02-21T02:42:13.759213938Z       /go/src/github.com/cilium/cilium/vendor/github.com/cilium/ebpf/syscalls.go:188 +0x808
2022-02-21T02:42:13.759220748Z   github.com/cilium/ebpf.init()
2022-02-21T02:42:13.759226898Z       /go/src/github.com/cilium/cilium/vendor/github.com/cilium/ebpf/syscalls.go:166 +0x7a8
2022-02-21T02:42:13.759233042Z   github.com/cilium/ebpf.init()
2022-02-21T02:42:13.759239198Z       /go/src/github.com/cilium/cilium/vendor/github.com/cilium/ebpf/syscalls.go:148 +0x748
2022-02-21T02:42:13.759245106Z   github.com/cilium/ebpf.init()
2022-02-21T02:42:13.759255136Z       /go/src/github.com/cilium/cilium/vendor/github.com/cilium/ebpf/syscalls.go:108 +0x6e8
2022-02-21T02:42:13.759261593Z   github.com/cilium/ebpf.init()
2022-02-21T02:42:13.759268165Z       /go/src/github.com/cilium/cilium/vendor/github.com/cilium/ebpf/syscalls.go:92 +0x688
2022-02-21T02:42:13.759274549Z   github.com/cilium/ebpf.init()
2022-02-21T02:42:13.759280966Z       /go/src/github.com/cilium/cilium/vendor/github.com/cilium/ebpf/syscalls.go:76 +0x628
2022-02-21T02:42:13.759286820Z   github.com/cilium/ebpf.init()
2022-02-21T02:42:13.759293199Z       /go/src/github.com/cilium/cilium/vendor/github.com/cilium/ebpf/syscalls.go:59 +0x5c8
2022-02-21T02:42:13.759299428Z   github.com/cilium/ebpf.init()
2022-02-21T02:42:13.759305593Z       /go/src/github.com/cilium/cilium/vendor/github.com/cilium/ebpf/syscalls.go:41 +0x568
2022-02-21T02:42:13.759311615Z   runtime.doInit()
2022-02-21T02:42:13.759318057Z       /usr/local/go/src/runtime/proc.go:6498 +0x122
2022-02-21T02:42:13.759324337Z   github.com/cilium/ebpf/internal/btf.init()
2022-02-21T02:42:13.759331415Z       /go/src/github.com/cilium/cilium/vendor/github.com/cilium/ebpf/internal/btf/btf.go:820 +0x1bc
```

### PR DESCRIPTION
**Pull request by aditighag**
Commit c9810bf7b2 introduced garbage collection for orphan entries
in the nat maps that led to concurrent accesses to the map.

The nat maps accessed via ct map construct are susceptible
to data races in asynchronously running goroutines upon
agent restart when endpoints are restored -

```
2022-02-21T02:42:13.757888057Z WARNING: DATA RACE
2022-02-21T02:42:13.757895621Z Write at 0x00c00081a830 by goroutine 360:
2022-02-21T02:42:13.757912783Z   github.com/cilium/cilium/pkg/bpf.(*Map).Close()
2022-02-21T02:42:13.757920669Z       /go/src/github.com/cilium/cilium/pkg/bpf/map_linux.go:581 +0x1c4
2022-02-21T02:42:13.757927597Z   github.com/cilium/cilium/pkg/maps/ctmap.doGC4·dwrap·4()
2022-02-21T02:42:13.757934561Z       /go/src/github.com/cilium/cilium/pkg/maps/ctmap/ctmap.go:422 +0x39
2022-02-21T02:42:13.757941184Z   github.com/cilium/cilium/pkg/maps/ctmap.doGC4()
2022-02-21T02:42:13.757947352Z       /go/src/github.com/cilium/cilium/pkg/maps/ctmap/ctmap.go:482 +0x4d5
2022-02-21T02:42:13.757953881Z   github.com/cilium/cilium/pkg/maps/ctmap.doGC()
2022-02-21T02:42:13.757960362Z       /go/src/github.com/cilium/cilium/pkg/maps/ctmap/ctmap.go:517 +0x17e
2022-02-21T02:42:13.757966185Z   github.com/cilium/cilium/pkg/maps/ctmap.GC()
2022-02-21T02:42:13.757972307Z       /go/src/github.com/cilium/cilium/pkg/maps/ctmap/ctmap.go:537 +0xc6
2022-02-21T02:42:13.757978599Z   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).garbageCollectConntrack()
2022-02-21T02:42:13.757986321Z       /go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:1034 +0x804
2022-02-21T02:42:13.757992160Z   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).scrubIPsInConntrackTableLocked()
2022-02-21T02:42:13.757998853Z       /go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:1039 +0x173
2022-02-21T02:42:13.758004601Z   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).scrubIPsInConntrackTable()
2022-02-21T02:42:13.758010701Z       /go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:1049 +0x44
2022-02-21T02:42:13.758016604Z   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).runPreCompilationSteps.func1()
2022-02-21T02:42:13.758022804Z       /go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:761 +0x132

2022-02-21T02:42:13.758034551Z Previous read at 0x00c00081a830 by goroutine 100:
2022-02-21T02:42:13.758040659Z   github.com/cilium/cilium/pkg/bpf.(*Map).DumpReliablyWithCallback()
2022-02-21T02:42:13.758046461Z       /go/src/github.com/cilium/cilium/pkg/bpf/map_linux.go:756 +0x804
2022-02-21T02:42:13.758053696Z   github.com/cilium/cilium/pkg/maps/nat.(*Map).DumpReliablyWithCallback()
2022-02-21T02:42:13.758059818Z       /go/src/github.com/cilium/cilium/pkg/maps/nat/nat.go:121 +0x7b7
2022-02-21T02:42:13.758065580Z   github.com/cilium/cilium/pkg/maps/ctmap.PurgeOrphanNATEntries()
2022-02-21T02:42:13.758072272Z       /go/src/github.com/cilium/cilium/pkg/maps/ctmap/ctmap.go:612 +0x790
2022-02-21T02:42:13.758078005Z   github.com/cilium/cilium/pkg/maps/ctmap/gc.runGC()
2022-02-21T02:42:13.758084196Z       /go/src/github.com/cilium/cilium/pkg/maps/ctmap/gc/gc.go:214 +0xb0c
2022-02-21T02:42:13.758090362Z   github.com/cilium/cilium/pkg/maps/ctmap/gc.runGC()
2022-02-21T02:42:13.758096712Z       /go/src/github.com/cilium/cilium/pkg/maps/ctmap/gc/gc.go:189 +0x6a5
2022-02-21T02:42:13.758103134Z   github.com/cilium/cilium/pkg/maps/ctmap/gc.runGC()
2022-02-21T02:42:13.758109338Z       /go/src/github.com/cilium/cilium/pkg/maps/ctmap/gc/gc.go:189 +0x6a5
2022-02-21T02:42:13.758127517Z   github.com/cilium/cilium/pkg/maps/ctmap/gc.Enable.func1()
2022-02-21T02:42:13.758135513Z       /go/src/github.com/cilium/cilium/pkg/maps/ctmap/gc/gc.go:90 +0x40a
2022-02-21T02:42:13.758141621Z   github.com/cilium/cilium/pkg/maps/ctmap/gc.runGC()
2022-02-21T02:42:13.758147715Z       /go/src/github.com/cilium/cilium/pkg/maps/ctmap/gc/gc.go:189 +0x6a5
2022-02-21T02:42:13.758153712Z   github.com/cilium/cilium/pkg/maps/ctmap/gc.Enable.func1()
2022-02-21T02:42:13.758160127Z       /go/src/github.com/cilium/cilium/pkg/maps/ctmap/gc/gc.go:90 +0x40a

```

Fixes: c9810bf7b2 ("ctmap: GC orphan SNAT entries")
Fixes: https://github.com/cilium/cilium/issues/18910

Signed-off-by: Aditi Ghag <aditi@cilium.io>

### CODE REVIEW

File: pkg/maps/ctmap/ctmap.go [hunk 1]
```diff
@@ -570,7 +584,9 @@ func PurgeOrphanNATEntries(ctMapTCP, ctMapAny *Map) *NatGCStats {
 
 	// Both CT maps should point to the same natMap, so use the first one
 	// to determine natMap
-	natMap := mapInfo[ctMapTCP.mapType].natMap
+	ctMap := mapInfo[ctMapTCP.mapType]
+	ctMap.natMapLock[ctMapTCP.mapType].Lock()
+	natMap := ctMap.natMap
 	if natMap == nil {
```
Comment (YutaroHayakawa) [inline pkg/maps/ctmap/ctmap.go:590]: Need unlocking in here?

Comment (YutaroHayakawa) [inline pkg/maps/ctmap/ctmap.go:592]: Need unlocking in here or simply defer?

**Review by YutaroHayakawa — CHANGES_REQUESTED** [review]
Almost seems good. Found one missing unlocking.

File: pkg/maps/ctmap/ctmap.go (hunk 1 above)
Comment (aditighag) [reply pkg/maps/ctmap/ctmap.go:592]: It's unlocked at L629 below. 

Edit : Oh, realized that there can be early returns so I've added a defer statement instead. Good spotting - Thanks!

**Comment by aditighag** [comment]
Simply moved the unlocking to a defer statement so no need to re-run the entire suite.

Screen Shot 2022-02-27 at 9 16 43 PM

File: pkg/maps/ctmap/ctmap.go (hunk 1 above)
Comment (YutaroHayakawa) [reply pkg/maps/ctmap/ctmap.go:592]: Now it's resolved.

**Review by YutaroHayakawa — APPROVED** [review]

**Comment by aditighag** [comment]
TIL : /test-race Thanks, Paul!

Travis has a successful run.
Sole failure in `test-net-next-race` is - https://github.com/cilium/cilium/issues/17604. I also double checked that there is no occurrence of "DATA RACE" in the sysdump. All other `test-*-race` jobs passed. Marking it as ready for merge.

**Comment by pchaigno** [comment]
The one failure for Travis CI was fixed by https://github.com/cilium/cilium/pull/18953 which this pull request doesn't include. I think it's safe to ignore.

**Comment by aditighag** [comment]
@YutaroHayakawa I've refactored the changes a bit so that we store only the corresponding lock for a global nat lock. This avoids retrieving the lock corresponding to a map type at the consumer site where we serialize accesses to the map. PTAL.

**Review by YutaroHayakawa — APPROVED** [review]
LGTM 👍

**Pull request merged**